src/main/scala/cognite/spark/v1/FlexibleDataModelCorePropertySyncRelation.scala

```diff
@@ -110,54 +110,120 @@ private[spark] class FlexibleDataModelCorePropertySyncRelation(
+      `with` = Map("sync" -> generateTableExpression(instanceType, matchNothingFilter)),
+      select = select)
+      .flatMap { sr =>
+        sr.nextCursor match {
+          case None =>
+            IO.raiseError(
+              new IllegalArgumentException("/sync response must have " +
+                "nextCursor"))
+          case Some(nextCursor) => IO.pure((nextCursor, cursors.isEmpty))
+        }
+      }
+      .redeemWith(
+        {
+          case e: CdpApiException if e.code == 400 && e.message.startsWith("Cursor has expired") =>
+            logger.warn("Cursor has expired, doing a full backfill")
+            generateFutureItemsCursor(None, instanceType, select)
+          case e => IO.raiseError(e)
+        },
+        IO.pure
+      )
+
+  private def generateTableExpression(
+      instanceType: InstanceType,
+      filters: FilterDefinition): TableExpression =
+    instanceType match {
+      case InstanceType.Edge =>
+        TableExpression(edges = Some(EdgeTableExpression(filter = Some(filters))))
+      case InstanceType.Node =>
+        TableExpression(nodes = Some(NodesTableExpression(filter = Some(filters))))
+    }
 
-    Seq(syncOut(instanceSyncRequest, selectedColumns))
+  private def fetchData(
+      isBackFill: Boolean,
+      cursors: Option[Map[String, String]],
+      `with`: Map[String, TableExpression],
+      select: Map[String, SelectExpression]): IO[ItemsWithCursor[InstanceDefinition]] = {
+    val response = if (isBackFill) {
+      client.instances.queryRequest(
+        InstanceQueryRequest(
+          `with` = `with`,
+          cursors = cursors,
+          select = select,
+          includeTyping = Some(true)
+        )
+      )
+    } else {
+      client.instances.syncRequest(
+        InstanceSyncRequest(
+          `with` = `with`,
+          cursors = cursors,
+          select = select,
+          includeTyping = Some(true)
+        )
+      )
+    }
+    response.map { qr =>
+      val itemDefinitions = qr.items.getOrElse(Map.empty).getOrElse("sync", Vector.empty)
+      val nextCursor = qr.nextCursor.getOrElse(Map.empty).get("sync")
+      ItemsWithCursor(itemDefinitions, nextCursor)
+    }
   }
-  // scalastyle:on cyclomatic.complexity
 
-  private def syncOut(syncRequest: InstanceSyncRequest, selectedProps: Array[String])(
+  private def syncOut(
+      futureItemsCursor: Option[String],
+      cursors: Option[Map[String, String]],
+      `with`: Map[String, TableExpression],
+      select: Map[String, SelectExpression],
+      selectedProps: Array[String])(
       implicit F: Async[IO]): Stream[IO, ProjectedFlexibleDataModelInstance] =
     Stream.eval {
-      syncWithCursor(syncRequest).map { sr =>
+      fetchData(futureItemsCursor.nonEmpty, cursors, `with`, select).map { sr =>
         val items = sr.items
         val nextCursor = sr.nextCursor
         val next =
           (nextCursor, items.nonEmpty && items.last.lastUpdatedTime < terminationTimeStamp) match {
```

Comment:
here we have a potential problem, if we decided to do a backfill but then decide to terminate based on this condition we may miss some items, even though terminationTimeStamp is set to now() and is close to futureCursor, but

- we don't limit /query endpoint and it may fetch recently updated items here, and they may also potentially appear at items.last position, as soon as we hit it we stop the iteration and assume nothing is left up to futureCursor, while we might have missed several remaining pages

So it seems what we need to do is always sync to the end in /query case as there's no contract of resuming /query iteration across runs

Reply:
applied a fix here